allocator.go

```diff
@@ -78,37 +78,102 @@ type Allocator struct {
+	AddNode()
+	UpdateNodeTags () bool
+	UpdateNodeResources() bool
+	RemoveNode () bool
+	AddRange ()
+	RemoveRange () bool
+}
+
+// AddNode Add a new Node
+func (nmap NodeMap) AddNode(id NodeID, tags []string, resources map[Resource]int64) {
+	nmap[id] = Node {
+		id:        id,
+		tags:      tags,
+		resources: resources,
+	}
+}
+
+// UpdateNodeTags Update tags of a Node
+func (nmap NodeMap) UpdateNodeTags(id NodeID, tags []string) bool {
+	if _, found := nmap[id]; found {
+		fmt.Println("Updating Node Tag ", id)
+		nmap[id]= Node {
+			id:        id,
+			tags:      tags,
+			resources: nmap[id].resources,
+		}
+		return true
+	}
+	fmt.Println("Node not found ", id)
+	return false
+}
+
+// UpdateNodeResources Update Resources of a Node
+func (nmap NodeMap) UpdateNodeResources(id NodeID, resources map[Resource]int64) bool {
+	if _, found := nmap[id]; found {
+		fmt.Println("Updating Node Resources ", id)
+		nmap[id]= Node {
+			id:        id,
+			tags:      nmap[id].tags,
+			resources: resources,
+		}
+		return true
+	}
+	fmt.Println("Node not found ", id)
+	return false
+}
+
+// RemoveNode Remove the node if its in the map
+func(nmap NodeMap) RemoveNode(n Node) bool{
+
+	if _, found := nmap[n.id]; found {
+		fmt.Println("Removing Node ", n.id)
+		delete(nmap,n.id)
+		return true
+	}
+	fmt.Println("Node not found ", n.id)
+	return true
+}
+
+func(rmap RangeMap) AddRange(id RangeID, rf int, tags []string, demands map[Resource]int64) {
+	rmap[id]= Range{
 		id:      id,
 		rf:      rf,
 		tags:    tags,
 		demands: demands,
 	}
 }
 
-// NewNode builds and returns nodes from the necessary parameters.
-func NewNode(id NodeID, tags []string, resources map[Resource]int64) Node {
-	return Node{
-		id:        id,
-		tags:      tags,
-		resources: resources,
+// RemoveRange Remove the range if its in the map
+func(rmap RangeMap) RemoveRange(r Range) bool{
+
+	if _, found := rmap[r.id]; found {
+		fmt.Println("Removing Range ", r.id)
+		delete(rmap,r.id)
+		return true
 	}
+	fmt.Println("Range not found ", r.id)
+	return true
 }
 
 // New builds, configures, and returns an allocator from the necessary parameters.
-func New(ranges []Range, nodes []Node, opts ...Option) *Allocator {
+func New(idRangeMap RangeMap, idNodeMap NodeMap, opts ...Option) *Allocator {
 	model := solver.NewModel("Lé-Allocator")
 	assignment := make(map[RangeID][]solver.IntVar)
 	// iterate over ranges, assign each rangeID a list of IV's sized r.rf.
 	// These will ultimately then read as: rangeID's replicas assigned to nodes [N.1, N.2,...N.RF]
-	for _, r := range ranges {
+	for _, r := range idRangeMap {
 		assignment[r.id] = make([]solver.IntVar, r.rf)
 		for j := range assignment[r.id] {
 			// constrain our IV's to live between [0, len(nodes) - 1].
 			assignment[r.id][j] = model.NewIntVarFromDomain(
-				solver.NewDomain(int64(nodes[0].id), int64(nodes[len(nodes)-1].id)),
+				solver.NewDomain(int64(idNodeMap[0].id), int64(idNodeMap[NodeID(len(idNodeMap)-1)].id)),
```

Comment:
Planning to delete all these usages?

Reply:
Why do we need to delete these?

Reply:
Ultimately it should end up in the `Allocate(...)` method.